File: sorting/O(n^2)/select-sort/bubble_sort.py
# see: https://www.digitalocean.com/community/tutorials/js-bubble-selection-insertion-sort
# 0(n^2) inefficient
# nested for loops gets largest value to end


def bubbleSort(arr):
    n = len(arr)

    # Traverse through all array elements
    for i in range(n):

        # Last i elements are already in place
        # n-i-1 at least stops inner loop short from iterating over already sorted item
        for j in range(0, n-i-1):

            # traverse the array from 0 to n-i-1
            # Swap if the element found is greater
            # than the next element
            if arr[j] > arr[j+1]:
                arr[j], arr[j+1] = arr[j+1], arr[j]


# Driver code to test above
arr = [31, 27, 28, 42, 13, 8, 11, 30, 17, 41, 15, 43, 1, 36, 9, 16, 20, 35, 48, 37, 7, 26, 34, 21,
       22, 6, 29, 32, 49, 10, 12, 19, 24, 38, 5, 14, 44, 40, 3, 50, 46, 25, 18, 33, 47, 4, 45, 39, 23, 2]

bubbleSort(arr)

print("Sorted array is:")
for i in range(len(arr)):
    print("%d" % arr[i]),

File: sorting/O(n^2)/select-sort/bubbleSort.js
import { unsortedArr } from "../../../utils/unsorted";

// see: https://www.digitalocean.com/community/tutorials/js-bubble-selection-insertion-sort
// 0(n^2) inefficient
// nested for loops gets largest value to end

// Bubble Sort
const bubbleSort = (arr) => {
  // helper swap function
  const swap = (array, a, b) => ([array[a], array[b]] = [array[b], array[a]]);

  for (let i = arr.length; i > 0; i--) {
    for (let j = 0; j < i - 1; j++) {
      if (arr[j] > arr[j + 1]) {
        swap(arr, j, j + 1);
      }
    }
  }

  return arr;
};

bubbleSort(unsortedArr);
